simple/cm_data.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) 2013-2015 Intel Corporation.  All rights reserved.
+ * Copyright (c) 2014-2016, Cisco Systems, Inc. All rights reserved.
+ *
+ * This software is available to you under the BSD license
+ * below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <string.h>
+#include <inttypes.h>
+#include <assert.h>
+
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_cm.h>
+
+#include "shared.h"
+
+static char *cm_data;
+static size_t cm_data_size;
+static struct fi_eq_cm_entry *entry;
+
+static int server_setup(void)
+{
+	size_t opt_size;
+	int ret;
+
+	ret = ft_start_server();
+	if (ret)
+		return ret;
+
+	opt_size = sizeof(opt_size);
+	return fi_getopt(&pep->fid, FI_OPT_ENDPOINT, FI_OPT_CM_DATA_SIZE,
+		&cm_data_size, &opt_size);
```

Comment:
> Good point, I'll close ofiwg/libfabric#1990. I'll update this PR to first create an active endpoint and retrieve the size.

@shefty I'm revisiting this and it isn't obvious to me what the usage model is for this functionality. On the client side there is already an active endpoint calling `fi_connect`, but the server side only has a passive endpoint and doesn't have any information about the active endpoint being requested. How is the server supposed to prepare for the data being received with the `FI_CONNREQ` event?

I think it would be a useful API addition to have `FI_OPT_CM_DATA_SIZE` return the maximum possible data size across all domains when used on a passive endpoint. It seems to me that without this the application would need call `fi_eq_read` and try again with a larger buffer if it gets `-FI_ETOOSMALL` back.

Thoughts?